Add some more typelabel support

akohlmey opened the pull request:
**Summary**

This pull request expands support for typelabels to a few more commonly used features in LAMMPS.

**Related Issue(s)**

 This expands on PR #4145 

**Author(s)**

Axel Kohlmeyer, Temple U

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

No known issues.

**Implementation Notes**

Type label support is added to the following features:
- neigh_modify exclude type
- fix shake/rattle

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

<!--Put any additional information here, attach relevant text or image files, and URLs to external sites (e.g. DOIs or webpages)-->

akohlmey committed:
use preferred comparison method

akohlmey committed:
add typelabel support to neigh_modify exclude type

akohlmey committed:
simplify parser for fix shake/rattle options and add typelabel support

akohlmey commented:
@jrgissing I would like to make an effort to have support for typelabels added to all features that are likely used in Simon's tutorial. While there should be one example showing that numeric types and typelabels can be mixed, all other examples using typelabels (and we should have a mix of numeric types and typelabels since typelabels just add overhead in some cases) should be fully supported.
I just added support for two more features used in my current text example.

akohlmey commented:
@lammps/core do you have any LAMMPS commands or styles where you use numeric types (for atoms, bonds, angles, etc) that have not yet been converted to support typelabels. See PR #4145 and this one for commands that have been converted.

I think for fix shake this also resulted in a cleaner parsing of the command arguments...

Here is an example that makes used of typelables throughout (and thus requires the changes in this pull request):
[typelabel-example.tar.gz](https://github.com/user-attachments/files/16366647/typelabel-example.tar.gz)

jrgissing commented on src/RIGID/fix_shake.cpp:
```diff
@@ -121,49 +121,40 @@ FixShake::FixShake(LAMMPS *lmp, int narg, char **arg) :
-    else if (mode == 'b') {
-      int i = utils::inumeric(FLERR,arg[next],false,lmp);
+    if ((strcmp(arg[next],"b") == 0) && (next+1 < narg)) {
+      int i = utils::expand_type_int(FLERR,arg[next+1],1,lmp);
```
I have been using enum values (e.g., Atom::ATOM) instead of numbers for the mode argument of expand_type_int

akohlmey replied:
Thanks for the reminder!
Laziness at work. Of course this should be using the enumerator. Will make the change ASAP.

jrgissing commented:
I purposely did not add support for rattle/shake in the previous PR, because I am not sure how the treat the case of a type label being 'a', 'b', 't' or 'm'.

akohlmey commented:
> I purposely did not add support for rattle/shake in the previous PR, because I am not sure how the treat the case of a type label being 'a', 'b', 't' or 'm'.

That is why I refactored the parser. The old version would loop over individual "words" and then set "mode" to the character from the previous iteration. This is rather difficult to understand and error prone.

Instead, the parser is now processing the "mode" character **and** the following word in one go and then just increment the argument index twice. Thus the check that was previously triggered is skipped.

Update:
I changed my test example to use "a" for hydrogen and "b" for oxygen and there is no problem with the revised fix shake parser.

akohlmey committed:
use enums instead of numbers.

jrgissing commented:
> > I purposely did not add support for rattle/shake in the previous PR, because I am not sure how the treat the case of a type label being 'a', 'b', 't' or 'm'.
> 
> That is why I refactored the parser. The old version would loop over individual "words" and then set "mode" to the character from the previous iteration. This is rather difficult to understand and error prone.
> 
> Instead, the parser is now processing the "mode" character **and** the following word in one go and then just increment the argument index twice. Thus the check that was previously triggered is skipped.
> 
> Update: I changed my test example to use "a" for hydrogen and "b" for oxygen and there is no problem with the revised fix shake parser.

I do not think this resolves all possible issues, because more than one (an unknown number of) types can be listed after each keyword

akohlmey commented:
> I do not think this resolves all possible issues, because more than one (an unknown number of) types can be listed after each keyword

Ok, so that is a problem. How about all typelabels would have to come with a "label:" or "l:" prefix, for example: b 1 2 l:HW-OW"?

akohlmey committed:
restore original fix shake code

akohlmey committed:
add check to detect whether there are overlaps between type labels and fix shake keywords

jrgissing commented:
How about we require quotes around all types for a certain 'type kind' when using type labels? This would be in line with the LAMMPS parsing style that quotes can enclose a single argument

akohlmey commented:
> How about we require quotes around all types for a certain 'type kind' when using type labels? This would be in line with the LAMMPS parsing style that quotes can enclose a single argument

After some deliberating, I am moving to a different strategy. I'm adding a test whether there is a type label conflicting with the 'b', 'a', 't', or 'm' keyword and then disable the checking for a type label (with a suitable warning message). I think the chance of conflict is very low and in case there is a conflict, there is always the option to change the type labels until there is no more conflict. Not perfect, but simpler and more consistent for the common case of no conflicts.

jrgissing commented:
> > How about we require quotes around all types for a certain 'type kind' when using type labels? This would be in line with the LAMMPS parsing style that quotes can enclose a single argument
> 
> After some deliberating, I am moving to a different strategy. I'm adding a test whether there is a type label conflicting with the 'b', 'a', 't', or 'm' keyword and then disable the checking for a type label (with a suitable warning message). I think the chance of conflict is very low and in case there is a conflict, there is always the option to change the type labels until there is no more conflict. Not perfect, but simpler and more consistent for the common case of no conflicts.

OK that sounds good

jrgissing commented on src/neighbor.cpp:
```diff
@@ -2705,8 +2705,8 @@ void Neighbor::modify_params(int narg, char **arg)
-        ex1_type[nex_type] = utils::inumeric(FLERR,arg[iarg+2],false,lmp);
-        ex2_type[nex_type] = utils::inumeric(FLERR,arg[iarg+3],false,lmp);
+        ex1_type[nex_type] = utils::expand_type_int(FLERR, arg[iarg+2], Atom::ATOM, lmp);
+        ex2_type[nex_type] = utils::expand_type_int(FLERR, arg[iarg+3], Atom::ATOM, lmp);
```
I would complain about not also supporting collection/type for neigh_modify, but that runs into the same parsing problem I had with pair_style tracker, where type labels are separated by commas

akohlmey replied:
I think those options are quite uncommon and thus not a high priority

akohlmey committed:
fix logic bug

akohlmey committed:
add conditional typelabel support

akohlmey committed:
update docs for added typelabel support

akohlmey commented:
> OK that sounds good

Just pushed the alternate implementation of type handling for fix shake/rattle

jrgissing commented on doc/src/fix_shake.rst:
```diff
@@ -27,9 +27,9 @@ Syntax
-       *b* values = one or more bond types
-       *a* values = one or more angle types
-       *t* values = one or more atom types
+       *b* values = one or more bond types (may use typelabels)
```
I think we have always used two words for 'type label'

akohlmey replied:
Ok. Made the change, also for fix shake docs.

jrgissing commented on doc/src/fix_shake.rst:
```diff
@@ -137,6 +137,13 @@ constrained (within a fudge factor of MASSDELTA specified in
 
+.. versionchanged:: TBD
+
+The types may be given as typelabels for as long as none of the
```
rewording suggestion: "The types may be given as type labels only if there is no atom, bond or angle type label of *b*, *a*, *t*, or *m* defined in the simulation.  In this case, type label support is disabled because the type arguments could be ambiguous."

sjplimp replied:
I suggest this altered 2nd sentence in Jake's suggestion, if it is correct:

If that is the case, type labels cannot be used as constraint values with these 2 fixes because those type labels would mistakenly be treated as a new category of constraint.  Use numeric types instead.

akohlmey replied:
Done

akohlmey committed:
make docs consistent with rest of manual and rephrase fix shake changes

akohlmey commented:
I am marking this as ready for review now since there doesn't seem to be any other common use cases for (numeric) types where we could support type labels.

jrgissing commented on doc/src/fix_shake.rst:
```diff
@@ -137,7 +137,17 @@ constrained (within a fudge factor of MASSDELTA specified in
+
+The types may be given as type labels *only* if there is no atom, bond,
+or angle type label named *b*, *a*, *t*, or *m* defined in the
+simulation.  In that is the case, type labels cannot be used as
```
In ...-> If...

akohlmey replied:
Fixed. Thanks.

akohlmey committed:
fix typo and rewrap
